### PR TITLE
docs: Changes to the Getting Started document

### PR DESCRIPTION
**Pull request by chipbuster**
This PR updates the "Getting Started" document with a more fleshed-out tutorial. Instead of running a single training step, it creates a model, generates noisy examples, and then attempts to learn the original model via examples. This shows new users what they have done, and how they can use the result of the training process.

### Potential Issues

According to BenchmarkTools, the old tutorial took a few milliseconds to run and consumed 18.5 KiB of memory, while the new one takes about 100ms to run and consumes 71.06 MiB of memory.

BenchmarkTools excludes compilation time, and in practice when I ran the two from the command line, the runtime appeared to be overwhelmingly dominated by the compilation time (approximately 20 seconds to complete each one), but the memory aspect may be of concern, depending on where users are expected to try this tutorial. I don't expect that this is an issue, since I couldn't even compile Flux on a VPS with 1GB RAM due to memory limits, but I don't know how this library is used. If 71MB of memory is a concern, I can attempt to cut down on the size of the training set at the cost of worse accuracy.

Also, in spite of the fact that I've written a SGD implementation or two, I'm actually not terribly familiar with the field of ML and am not particularly fluent with how things are usually talked about/described. If anyone would like wording or vocabulary changes to make things more idiomatic, I'd be happy to change those.

### CODE REVIEW

File: getting_started.md [hunk 1]
```diff
@@ -52,74 +52,84 @@ using Flux
...
+Training a model is the process of computing the gradients with respect to the parameters for each input in the data. At every step, the optimiser updates all of the parameters until it finds a good value for them. This process can be written as a loop: we iterate over the examples in `x_train` and `y_train` and update the model for each example.
 
-You can execute the training process of your model as follows:
+To indicate that we want all derivatives of `W_learned` and `b_learned`, we write `ps = params([W, b])`. This is a convenience function that Flux provides so that we don't have to explicitly list every gradient we want. Check out the section on [Taking Gradients](https://fluxml.ai/Flux.jl/stable/models/basics/#Taking-Gradients) if you want to learn more about how this works.
```
Comment (DhairyaLGandhi) [inline getting_started.md:122]:
```suggestion
To indicate that we want all derivatives of `W_learned` and `b_learned`, we write `ps = params(W, b)`. This is a convenience function that Flux provides so that we don't have to explicitly list every gradient we want. Check out the section on [Taking Gradients](https://fluxml.ai/Flux.jl/stable/models/basics/#Taking-Gradients) if you want to learn more about how this works.
```

File: getting_started.md [hunk 2]
```diff
@@ -128,79 +138,108 @@ end
...
+>**Note:** With this pattern, it is easy to add more complex learning routines that make use of control flow, distributed compute, scheduling optimisation, etc. Note that the pattern above is a simple julia *for loop* but it could also be replaced with a *while loop*.
 
 <br>
 
-Flux enables you to execute the same process with the [Flux.train!](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1) function. It executes one training step, and you can put the `Flux.train!` function inside a *for loop* to execute more training steps. For more information on training a model in Flux, see [Training](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1). 
+Flux lets you do the same process with the [Flux.train!](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1) function. `train!` executes a single training step, and you can also put it inside a for-loop to execute multiple training steps. For more information on training a model in Flux, see [Training](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1).
```
Comment (DhairyaLGandhi) [inline getting_started.md:145]: `train!` doesn't do a single training step, it iterates over all the input data, effectively an epoch.

File: getting_started.md [hunk 3]
```diff
@@ -128,79 +138,108 @@ end
...
+>**Note:** With this pattern, it is easy to add more complex learning routines that make use of control flow, distributed compute, scheduling optimisation, etc. Note that the pattern above is a simple julia *for loop* but it could also be replaced with a *while loop*.
 
 <br>
 
-Flux enables you to execute the same process with the [Flux.train!](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1) function. It executes one training step, and you can put the `Flux.train!` function inside a *for loop* to execute more training steps. For more information on training a model in Flux, see [Training](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1). 
+Flux lets you do the same process with the [Flux.train!](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1) function. `train!` executes a single training step, and you can also put it inside a for-loop to execute multiple training steps. For more information on training a model in Flux, see [Training](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1). 
 
 ```julia
-Flux.train!(loss, params(model), data, opt)
+for d in train_data
```
Comment (DhairyaLGandhi) [inline getting_started.md:148]: See comment above

**Comment by DhairyaLGandhi** [comment]
Thanks for taking a look, generally pretty good ideas! It is considered good practice to remove the compilation time while benchmarking. We probably want to keep things as simple as possible for a getting started guide. I do believe we have some guides like in the PR too.

File: getting_started.md (hunk 1 above)
Comment (chipbuster) [reply getting_started.md:122]: D'oh. Thanks for the catch.

**Comment by chipbuster** [comment]
Definitely agreed about the simplicity. I think the older document did a great job in being as simple as possible with the training of the model, but I think it was potentially unclear about what exactly you were training or what you could do with it afterwards.

If there are any areas of this PR that you think should be cut or moved elsewhere, please do point them out! Striking the right balance between enough detail and too much detail is always a challenge.

File: getting_started.md [hunk 4]
```diff
@@ -128,83 +138,112 @@ end
...
-### Step 7: Run the script
+### Step 7: Examine the Results
+
+The training loop we ran modified `W_learned` and `b_learned` to be closer to the values used to generate the training data (`W_actual` and `b_actual`). We can see how well we did by printing out the difference between the learned and actual matrices.
+
+```julia
+println("The value of W_learned is:")
+display(W_learned)
+print("The largest element-wise difference between W_learned and W_actual is ")
+println(maximum(abs.(W_learned - W_actual))
```
Comment (ToucheSir) [inline getting_started.md:165]: Instead of writing about what each value represents in prose, I think a simple `@show` would do the trick:
```suggestion
@show W_learned
@show maximum(abs, W_learned .- W_actual)
```

File: getting_started.md [hunk 5]
```diff
@@ -52,74 +52,84 @@ using Flux
...
+b_actual = [- 1.0; -2.0]
+actual_model(x) = W_actual * x + b_actual
 ```
 <br>
 
+Next, we generate our training data by passing random vectors into the true model. We'll also add Gaussian noise using `randn()` so that it's not *too* easy for Flux to figure out the model.
+
+```julia
+x_train = [ 5 .* rand(5) for _ in 1:10_000 ]
+y_train = [ actual_model(x) + 0.2 .* randn(2) for x in x_train
```
Comment (ToucheSir) [inline getting_started.md:70]:
```suggestion
y_train = [ actual_model(x) + 0.2 .* randn(2) for x in x_train]
```
I also think it would help to note what structure `{x,y}_train` actually have. I don't think it's immediately clear that we're dealing with vectors of arrays here, especially since in practice each batch would be a 2D+ array and the outer wrapper "collection" (vector, tuple, dataloader, etc.) is 100% irrelevant.

File: getting_started.md [hunk 6]
```diff
@@ -52,74 +52,84 @@ using Flux
...
+In most real machine learning scenarios, you would not have access to `actual_model`, and would just get `x_train` and `y_train` as input data.
+
 ### Step 3: Define your model
 
-Define a simple regression model by defining the following function:
+Next, we define the model we want to use to learn the data. We'll use the same form that we used for our training data:
 
 ```julia
-model(x) = W*x .+ b
+training_model(x) = W_learned*x + b_learned
```
Comment (ToucheSir) [inline getting_started.md:80]:
```suggestion
training_model(x) = W_learned*x .+ b_learned
```
`.+` is good broadcasting hygiene (`+` on arrays calls it behind the scenes). I'd also lean towards calling `training_model` just `model` and renaming `actual_model` to better represent its role as a ground truth function (side note: is "true model" an oxymoron?)

File: getting_started.md [hunk 7]
```diff
@@ -52,74 +52,84 @@ using Flux
...
+We can now execute the training procedure for our model:
 
 ```julia
-data = [(x, y)]
-ps = params(W, b)
+train_data = zip(x_train, y_train)
+ps = params(W_learned, b_learned)
 
-for d in data
+for d in train_data
```
Comment (ToucheSir) [inline getting_started.md:130]:
```suggestion
for (x, y) in train_data
```
and passing them into loss separately, would save us from having to slap down a comment below. I don't recall why the splat was added, but IMO it's wholly unnecessary now.

File: getting_started.md [hunk 8]
```diff
@@ -128,83 +138,112 @@ end
...
+>**Note:** With this pattern, it is easy to add more complex learning routines that make use of control flow, distributed compute, scheduling optimisations, etc. Note that the pattern above is a simple Julia *for loop* but it could also be replaced with a *while loop*.
 
 <br>
 
-Flux enables you to execute the same process with the [Flux.train!](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1) function. It executes one training step, and you can put the `Flux.train!` function inside a *for loop* to execute more training steps. For more information on training a model in Flux, see [Training](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1). 
+While writing your own loop is powerful, sometimes you just want to do the simple thing without writing too much code. Flux lets you do this with the [Flux.train!](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1) function, which computes a gradient and updates the model for *every* example in a set (running what is often referred to as a *training epoch*). In our case, we could have replaced the entire loop with the following statement:
```
Comment (ToucheSir) [inline getting_started.md:145]:
```suggestion
While writing your own loop is powerful, sometimes you just want to do the simple thing without writing too much code. Flux lets you do this with [Flux.train!](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1), which runs one training epoch over a dataset. `Flux.train!` computes gradients and updates model parameters for every sample or batch of samples. In our case, we could have replaced the above loop with the following statement:
```

**Review by ToucheSir — CHANGES_REQUESTED** [review]
Thanks @chipbuster, I've done a first pass review. Feel free to ping if you have any questions.

File: getting_started.md (hunk 6 above)
Comment (chipbuster) [reply getting_started.md:80]: I was suspicious of this since broadcasting is partially what allowed #101 to slip through in the first place (without broadcasting, the dimensions would have failed to match), but you're right in that it's probably better to set a good example here than it is to guard against a one-off case.

Perhaps `model` and `ground_truth`?

File: getting_started.md (hunk 5 above)
Comment (chipbuster) [reply getting_started.md:70]: Would it be preferable to switch this to a 2D array and then use `data = zip(eachcol(x_train), eachcol(y_train))` further down? That was one of my initial thoughts.

Comment (chipbuster) [reply getting_started.md:70]: > switch this to a 2D array

Nope, this creates awkwardness when generating examples. I'll just clarify this in prose.

File: getting_started.md (hunk 6 above)
Comment (chipbuster) [reply getting_started.md]: Perhaps there is no true model, only a most-useful model.

File: getting_started.md [hunk 9]
```diff
@@ -52,25 +52,41 @@ using Flux
...
-For this example, create some random train data `x` and `y` arrays:
+### Step 2: Create the training data
+First, we'll write a function that generates our "true" data. We'll use to use Flux to recover `W_truth` and `b_truth` by looking only at examples of the `ground_truth` function.
 
 ```julia
-x = rand(5)
-y = rand(2) 
+W_truth = [1 2 3 4 5;
+            5 4 3 2 1]
+b_truth = [- 1.0; -2.0]
```
Comment (ToucheSir) [inline getting_started.md:61]:
```suggestion
b_truth = [-1.0; -2.0]
```

File: getting_started.md [hunk 10]
```diff
@@ -80,46 +96,45 @@ b = rand(2)
...
+A loss function evaluates a machine learning model's performance. In other words, it measures how far the model is from its target prediction. Flux lets you define your own custom loss function, or you can use one of the [Loss Functions](https://fluxml.ai/Flux.jl/stable/training/training/#Loss-Functions-1) that Flux provides. 
 
-For this example, define a custom loss function:
+For this example, we'll define a loss function that measures the squared distance from the predicted output to the actual output:
 
 ```julia
 function loss(x, y)
   ŷ = model(x)
-  sum((y .- ŷ).^2)
+  sum((y - ŷ).^2)
```
Comment (ToucheSir) [inline getting_started.md:106]:
```suggestion
  sum((y .- ŷ).^2)
```
I understand the want to avoid strange behaviour with broadcasting differently sized operands, but we should be consistent and removing the dot will break broadcast fusion between `-` and `^` (inconsequential here, but a big deal for more expensive forward passes).

File: getting_started.md [hunk 11]
```diff
@@ -52,25 +52,41 @@ using Flux
...
+### Step 2: Create the training data
+First, we'll write a function that generates our "true" data. We'll use to use Flux to recover `W_truth` and `b_truth` by looking only at examples of the `ground_truth` function.
 
 ```julia
-x = rand(5)
-y = rand(2) 
+W_truth = [1 2 3 4 5;
+            5 4 3 2 1]
+b_truth = [- 1.0; -2.0]
+ground_truth(x) = W_truth * x .+ b_truth
```
Comment (ToucheSir) [inline getting_started.md:62]:
```suggestion
ground_truth(x) = W_truth*x .+ b_truth
```
To match the function below

File: getting_started.md [hunk 12]
```diff
@@ -52,25 +52,41 @@ using Flux
...
+    only the training examples.
+
 ### Step 3: Define your model
 
-Define a simple regression model by defining the following function:
+Next, we define the model we want to use to learn the data. We'll use the same form that we used for our training data:
 
 ```julia
-model(x) = W*x .+ b
+model(x) = W*x + b
```
Comment (ToucheSir) [inline getting_started.md:85]:
```suggestion
model(x) = W*x .+ b
```
To match the function above

**Review by ToucheSir — COMMENTED** [review]
Thanks, just some small nits and we're good to go :)

File: getting_started.md (hunk 11 above)
Comment (chipbuster) [reply getting_started.md:62]: Actually, I just totally forgot about those. Thanks for the catch.

File: getting_started.md (hunk 10 above)
Comment (chipbuster) [reply getting_started.md:106]: Actually, I just totally forgot about those. Thanks for the catch.

**Comment by chipbuster** [comment]
And I need to do a pass to make the bulk code at the end match these changes. Might take me a minute since I'm not on my normal setup right now.

**Comment by chipbuster** [comment]
@ToucheSir Sorry for the delay.

**Comment by ToucheSir** [comment]
No need to apologize, this is a pretty fast turnaround for a docs PR :). Let me make a couple tweaks and we'll be good to merge.

File: getting_started.md [hunk 13]
```diff
@@ -128,83 +143,91 @@ end
...
+@show maximum(abs, W .- W_truth)
 ```
 
 <br>
 
 
 ## What's next
 
-Congratulations! You have created your first model and ran a training step using Flux. Now, you can continue exploring Flux's capabilities:
+Congratulations! You have created written and trained a model using Flux. Now, you can continue exploring Flux's capabilities:
```
Comment (ToucheSir) [inline getting_started.md:226]:
```suggestion
Congratulations! You have created and trained a model using Flux. Now, you can continue exploring Flux's capabilities:
```

File: getting_started.md [hunk 14]
```diff
@@ -128,83 +143,91 @@ end
...
+end
+opt = Descent(0.01)
+train_data = zip(x_train, y_train)
+ps = params(W, b)
 
-# Training process
-for d in data
+# Execute a training epoch
+for (x,y) in train_data
   gs = Flux.gradient(ps) do
```
Comment (ToucheSir) [inline getting_started.md:207]:
```suggestion
  gs = gradient(ps) do
```

**Review by ToucheSir — APPROVED** [review]
Thanks @chipbuster!
